### PR TITLE
[Sharktank] Llm Task Scheduler

### PR DESCRIPTION
**Pull request by stbaione**
## Description

- Introduces a `Scheduler` into the `vmfb-runner`
- Adds a `submit + run` flow to the `Batcher`
    - Each request is given a single `LlmTaskInput`, which gets added to the `Scheduler`, with a `count` variable.
    - This `count` var allows the scheduler to determine how long each request should be tracked. For example, the `decode_scheduler` will track each task for at most `steps` number of schedules.
 - The `Decoder` submits the tasks to the `Batcher`, then returns the results from calling `run`
 - Currently, keeping the standalone prefill/decode functions for the other components using `Batcher` (i.e. perplexity). Would need to think of a way to handle the fact that PPL needs all of the raw logits, not just the selections, but that should be possible to do. Maybe by moving the `selection` logic back outside of `Batcher`, and just providing it a callback for where it should submit the logits + indices.
 - Splits components into `llm_scheduler.py`, `llm_task.py` and `llm_utils.py`, because the file was getting complex

### CODE REVIEW

File: sharktank/sharktank/utils/llm_task.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+class LlmTaskInput:
+    task_id: str
+    tokens: List[int]
+    seq_len: int
+    pages: List[int]
+
+    start_position: Optional[int] = None
+
+
+class LlmTask(ABC):
```
Comment (rsuderman) [inline sharktank/sharktank/utils/llm_task.py:30]: I would move this to the scheduler and generalize it.  Most of the basics are pretty use case vague and it would be easy to write some tests on the scheduler + task

File: sharktank/sharktank/utils/llm_utils.py [hunk 2]
```diff
@@ -538,6 +391,83 @@ def decode(
...
+
+        for i, request in enumerate(requests):
+            task_input = LlmTaskInput(
+                task_id=f"req_{i}",
+                tokens=request,
+                seq_len=len(request),
+                pages=page_ids[i],
+            )
+            # Submit prefill task
+            self._prefill_scheduler.schedule_task(task_input, 1)
```
Comment (rsuderman) [inline sharktank/sharktank/utils/llm_utils.py:410]: This feels wrong - why are you submitting to both prefill and decode?

File: sharktank/sharktank/utils/llm_utils.py [hunk 3]
```diff
@@ -538,6 +391,83 @@ def decode(
...
+            if not self._decode_scheduler.has_pending_tasks():
+                break
+            task_inputs = self._decode_scheduler.get_scheduled_tasks()
+            decode_task = DecodeTask(
+                invocation_fn=self._instance.decode,
+                llm_task_inputs=task_inputs,
+                batch_size=self._decode_bs,
+                block_stride=self._block_stride,
+            )
+            logits, indices = decode_task.run(*self._cache)
```
Comment (rsuderman) [inline sharktank/sharktank/utils/llm_utils.py:458]: This looks wrong.  You make a task then call run on it then  call the scheduler? You should be making a job and submitting it. The scheduler should basically have a `schedule` command for each task (a task being a separate request). Then a call for competion. Something is off on this.

File: sharktank/sharktank/utils/llm_utils.py [hunk 4]
```diff
@@ -538,6 +391,83 @@ def decode(
...
+            last = selection_fn(logits, indices, [0] * len(task_inputs))
+            for task_input, token in zip(task_inputs, last):
+                selections_map[task_input.task_id].append(token)
+                self._decode_scheduler.on_task_complete(
```
Comment (rsuderman) [inline sharktank/sharktank/utils/llm_utils.py:462]: This on task completion callback is wrong. You should submit then process on the returned results.

File: sharktank/sharktank/utils/llm_utils.py [hunk 5]
```diff
@@ -538,6 +391,83 @@ def decode(
...
+                    task_input.task_id,
+                    token,
+                )
+                if token == eos_token:
+                    self._decode_scheduler.remove_task(task_input.task_id)
```
Comment (rsuderman) [inline sharktank/sharktank/utils/llm_utils.py:467]: Schedulers should take requests, process them, then be done. You shouldn't be externally managing.

File: sharktank/sharktank/utils/llm_scheduler.py
Comment (rsuderman) [inline sharktank/sharktank/utils/llm_scheduler.py:1]: Make enough parts in scheduler and write some tests for it. I think you will understand the issues with your API if you attempt to use it standalone vs retrofitting it into the old process.

**Review by rsuderman — CHANGES_REQUESTED** [review]
